core/src/main/scala/org/bitcoins/core/crypto/TxSigComponent.scala

```diff
@@ -44,6 +46,130 @@ sealed abstract class TxSigComponent {
 
 object TxSigComponent {
 
+  def apply(
+      inputInfo: InputInfo,
+      unsignedTx: Transaction,
+      flags: Seq[ScriptFlag] = Policy.standardFlags): TxSigComponent = {
+    inputInfo match {
+      case segwit: SegwitV0NativeInputInfo =>
+        fromWitnessInput(segwit, unsignedTx, flags)
+      case unassigned: UnassignedSegwitNativeInputInfo =>
+        fromWitnessInput(unassigned, unsignedTx, flags)
+      case p2sh: P2SHInputInfo =>
+        fromP2SHInput(p2sh, unsignedTx, flags)
+      case raw: RawInputInfo =>
+        fromRawInput(raw, unsignedTx, flags)
+    }
+  }
+
+  def fromWitnessInput(
+      inputInfo: SegwitV0NativeInputInfo,
+      unsignedTx: Transaction,
+      flags: Seq[ScriptFlag]): TxSigComponent = {
+    val idx = TxUtil.inputIndex(inputInfo, unsignedTx)
+    val wtx = {
+      val unsignedWtx = WitnessTransaction.toWitnessTx(unsignedTx)
+      unsignedWtx.witness.witnesses(idx) match {
+        // Only set the witness if we don't already have one
+        case EmptyScriptWitness =>
+          InputInfo.getScriptWitness(inputInfo) match {
+            case None =>
+              unsignedWtx
+            case Some(scriptWitness) =>
+              unsignedWtx.updateWitness(idx, scriptWitness)
+          }
+        case _: ScriptWitnessV0 =>
+          unsignedWtx
+      }
+    }
+
+    WitnessTxSigComponent(wtx, UInt32(idx), inputInfo.output, flags)
+  }
+
+  def fromWitnessInput(
+      inputInfo: UnassignedSegwitNativeInputInfo,
+      unsignedTx: Transaction,
```

Comment:
Doesn't this have to be a `WitnessTransaction`?

Reply:
No, because sometimes we'll have unsigned txs that are considered `BaseTransaction`s. This shouldn't cause any problems because in the these functions we always make it a `WitnessTransaction` and set the witness data if needed.